src/glib/mine/svm.h

```diff
@@ -13,376 +13,640 @@
+                ++SV;
+            }
+            SVIndices.SetVal(Idx, svm_model->sv_indices[Idx]);
+            for (int cIdx = 0; cIdx < svm_model->nr_class - 1; cIdx++){
+              Coef.PutXY(cIdx, Idx, svm_model->sv_coef[cIdx][Idx]);
             }
+        }        
+        for (int Idx = 0; Idx < svm_model->nr_class * (svm_model->nr_class - 1)/2; Idx++){
+            Rho.SetVal(Idx, svm_model->rho[Idx]);
+        }          
+        for (int Idx = 0; Idx < svm_model->nr_class; Idx++){
+          Label.SetVal(Idx, svm_model->label[Idx]);
+          nSV.SetVal(Idx, svm_model->nSV[Idx]);
         }
-        x_space[N].index = -1;
-        ++N;
-        svm_problem.x[Idx] = &x_space[prevN];
+        
+        this->WgtV = WgtV;
+        this->Bias = Bias;
+        this->NClass = svm_model->nr_class;
+        this->NSV = svm_model->l;
+        this->SV = SVs;
+        this->Coef = Coef;
+        this->Rho = Rho;
+        this->ProbA = ProbA;
+        this->ProbB = ProbB;
+        this->SVIndices = SVIndices;
+        this->Label = Label;
+        this->nSV = nSV;
+        this->Free = svm_model->free_sv;
+
+        svm_free_and_destroy_model(&svm_model);
+        svm_destroy_param(&svm_parameter);
+        free(svm_problem.y);
+        free(svm_problem.x);
+        free(x_space);
+        free(svm_model);
     }
 
-    const char* error_msg = svm_check_parameter(&svm_problem, &svm_parameter);
-    EAssertR(error_msg == NULL, error_msg);
-
-    // Learn the model
-    svm_model_t* svm_model = svm_train(&svm_problem, &svm_parameter, DebugNotify, ErrorNotify);
-
-    // Make sure the WgtV is non-null, i.e., in case w=0 set WgtV to a vector
-    // composed of a sufficient number of zeros (e.g. [0, 0, ..., 0]).
-    TFltV WgtV(DimN);
-    TFlt Bias = -svm_model->rho[0]; // LIBSVM does w*x-b, while we do w*x+b; thus the sign flip
-    EAssertR(TLinAlg::Norm(WgtV) == 0.0, "Expected a zero weight vector.");
-    for (int Idx = 0; Idx < svm_model->l; ++Idx) {
-        svm_node_t* SV = svm_model->SV[Idx];
-        while (SV->index != -1) {
-            WgtV[SV->index - 1] += svm_model->sv_coef[0][Idx] * SV->value;
-            ++SV;
+    // Use LIBSVM for C-Support Vector Classification
+    void LibSvmSolveClassify(const TFltVV& VecV, const TFltV& TargetV,
+            PNotify DebugNotify, PNotify ErrorNotify) {
+      
+        // load training parameters
+        svm_parameter_t svm_parameter = this->Param.GetParamStruct();
+
+        // Asserts for input arguments
+        EAssertR(this->Param.Cost > 0.0, "Cost parameter has to be positive.");
+
+        const int DimN = VecV.GetXDim(); // Number of features
+        const int AllN = VecV.GetYDim(); // Number of examples
+
+        EAssertR(TargetV.Len() == AllN, "Dimension mismatch.");
+
+        svm_problem_t svm_problem;
+        svm_problem.l = AllN;
+        svm_problem.y = (double *)malloc(AllN*sizeof(double));
+
+        svm_problem.x = (svm_node_t **)malloc(AllN*sizeof(svm_node_t *));
```

Comment:
This looks same as code above, can we move it to a separate helper function that copies data from input to `svm_problem.x`?